### PR TITLE
feat: healthcheck, node and hooks server

### PR DESCRIPTION
**Pull request by YuriBuerov**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: packages/sdk/src/middleware/server.ts [hunk 1]
```diff
@@ -165,6 +165,14 @@ export const createServer = async ({
 
 	fastify.decorateRequest('ctx', null);
 
+	fastify.route({
+		method: 'GET',
+		url: '/',
```
Comment (StarpTech) [inline packages/sdk/src/middleware/server.ts:170]: Let's call it "/health" to make it clear.

File: pkg/node/node.go [hunk 2]
```diff
@@ -411,8 +411,18 @@ func (n *Node) startServer(nodeConfig WunderNodeConfig) error {
 	}()
 
 	router.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (StarpTech) [inline pkg/node/node.go:413]: Here too call it "/health"

File: pkg/node/node.go [hunk 3]
```diff
@@ -411,8 +411,18 @@ func (n *Node) startServer(nodeConfig WunderNodeConfig) error {
...
-		_, _ = w.Write([]byte("WunderNode Status: OK\n"))
-		_, _ = w.Write([]byte(fmt.Sprintf("BuildInfo: %+v\n", n.info)))
+		resp, err := http.DefaultClient.Get(serverUrl + "/")
```
Comment (StarpTech) [inline pkg/node/node.go:414]: Please ensure proper max timeouts.

File: pkg/node/node.go [hunk 4]
```diff
@@ -411,8 +411,18 @@ func (n *Node) startServer(nodeConfig WunderNodeConfig) error {
...
+		if err != nil || resp.StatusCode != 200 {
+			serverStatus = "DEAD"
+		}
+
+		health := HealthCheck{
+			ServerStatus: serverStatus,
+			NodeStatus:   "OK",
+			BuildInfo:    n.info,
+		}
+		_ = json.NewEncoder(w).Encode(health)
```
Comment (StarpTech) [inline pkg/node/node.go:425]: The endpoint has to return with a non-200 status code when the server is unhealthy.

File: pkg/node/node.go (hunk 3 above)
Comment (StarpTech) [inline pkg/node/node.go:414]: This check should only happen when wunderctl manage the server process "wunderctl start"

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: cli/commands/node.go [hunk 5]
```diff
@@ -78,7 +78,24 @@ func NewWunderGraphNode(ctx context.Context) (*node.Node, error) {
...
+
+type Option func(options *options)
+
+func WithHooksServerHealthCheck() Option {
+	return func(options *options) {
+		options.hooksServerHealthCheck = true
+	}
+}
+
+func StartWunderGraphNode(n *node.Node, stop func(), opts ...Option) error {
```
Comment (StarpTech) [inline cli/commands/node.go:93]: Could you convert "stop" also to an option?

File: pkg/node/node.go [hunk 6]
```diff
@@ -410,11 +424,21 @@ func (n *Node) startServer(nodeConfig WunderNodeConfig) error {
...
-	router.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		resp, err := http.DefaultClient.Get(serverUrl + "/")
-		serverStatus := "OK"
-		if err != nil || resp.StatusCode != 200 {
-			serverStatus = "DEAD"
+	router.Handle("/health", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		serverStatus := "SKIP"
+		if n.options.hooksServerHealthCheck {
+			c := http.DefaultClient
+			c.Timeout = 30 * time.Second
```
Comment (StarpTech) [inline pkg/node/node.go]: I'd suggest creating a custom HTTP client outside of the handler.

File: pkg/node/node.go [hunk 7]
```diff
@@ -410,11 +424,21 @@ func (n *Node) startServer(nodeConfig WunderNodeConfig) error {
...
+			if n.options.healthCheckTimeout != nil {
+				c.Timeout = *n.options.healthCheckTimeout
+			}
+
+			resp, err := c.Get(serverUrl + "/health")
```
Comment (StarpTech) [inline pkg/node/node.go:431]: Use

```
url.JoinPath("http://host/foo", "bar/")
```

File: pkg/node/node.go (hunk 6 above)
Comment (StarpTech) [inline pkg/node/node.go]: The default should come from the cli flag default value.

File: cli/commands/start.go [hunk 8]
```diff
@@ -71,4 +72,5 @@ func init() {
 	startCmd.Flags().BoolVar(&enableIntrospection, "enable-introspection", false, "enables GraphQL introspection on /%api%/%main%/graphql")
 	startCmd.Flags().BoolVar(&disableForceHttpsRedirects, "disable-force-https-redirects", false, "disables authentication to enforce https redirects")
 	startCmd.Flags().IntVar(&shutdownAfterIdle, "shutdown-after-idle", 0, "shuts down the server after given seconds in idle when no requests have been served")
+	startCmd.Flags().IntVar(&healthCheckTimeout, "healthcheck-timeout", 0, "healthcheck timeout in seconds")
```
Comment (StarpTech) [inline cli/commands/start.go:75]: You can define a default here.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/node/node.go (hunk 7 above)
Comment (spetrunin) [reply pkg/node/node.go:431]: It will require go 1.19

Comment (StarpTech) [reply pkg/node/node.go:431]: :cry:

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**